src/consensus/blockstore.rs

```diff
@@ -105,230 +81,121 @@ impl Blockstore {
         shred: Shred,
         check_equivocation: bool,
     ) -> Option<(Slot, BlockInfo)> {
-        if check_equivocation && self.equivocated_slots.contains(&shred.payload().slot) {
-            debug!("recevied shred from equivocating leader, not adding to blockstore");
-            return None;
-        }
         let slot = shred.payload().slot;
-        let slice = shred.payload().slice_index;
-        let index = shred.payload().index_in_slice;
-        let is_last_slice = shred.payload().is_last_slice;
-        let slice_shreds = self.shreds.entry((slot, slice)).or_default();
-
-        // check Merkle root and signature
-        let leader_pk = &self.epoch_info.leader(slot).pubkey;
-        let cached_merkle_root = self.merkle_root_cache.get(&(slot, slice));
-        if !shred.verify(leader_pk, cached_merkle_root) {
-            debug!("dropping invalid shred");
-            return None;
-        } else if cached_merkle_root.is_none() {
-            self.merkle_root_cache
-                .insert((slot, slice), shred.merkle_root);
-        } else if cached_merkle_root != Some(&shred.merkle_root) {
-            if self.equivocated_slots.insert(slot) {
-                warn!("shreds show leader equivocation in slot {slot}");
+        let leader_pk = self.epoch_info.leader(slot).pubkey;
+        let event = self
+            .slot_data_mut(slot)
+            .add_shred(shred.clone(), check_equivocation, leader_pk)
+            .await;
+
+        match event.clone()? {
```

Comment:
```suggestion
        let event = event?;
        match &event {
```

I think then we can remove the `unwrap`s below maybe?  I haven't tested the theory though.

Reply:
does not work, because `send()` takes the event by value and you cannot move `event` while it's still borrowed